[IMP] web_editor: change image cropping from modal to inline widget

sdegueldre opened the pull request:
Previously, cropping an image was done inside of a modal window, turning
it into an inline widget makes the editor feel more integrated allows
you to better see the result of cropping directly in the page, it's also
less intrusive to the user's edition workflow.

task-2192755

sdegueldre commented:
@qsm-odoo let me know if you would like me to add a move commit before this one so that the `crop_dialog` history can be followed from `image_crop_widget.js`

qsm-odoo commented:
> @qsm-odoo let me know if you would like me to add a move commit before this one so that the `crop_dialog` history can be followed from `image_crop_widget.js`

Always.

qsm-odoo commented:
@sdegueldre Ping me when you judge the commits are clean for a first small review before continuing :+1:

seb-odoo commented:
Code owners sent me here. Changes in the base files looks ok, since its just removing a method used only there.

Not sure about the state and the scope of the PR, but I find strange the change to guess original based on URL. Is that for being backward compatible? Don't we have a more robust way to keep the information about original in the future?

That said I feel like this is getting out of scope of what I should check here, so as long as no major issue is introduced in the base file, I'll let you guys handle it as you think is best.

sdegueldre commented:
> Not sure about the state and the scope of the PR, but I find strange the change to guess original based on URL. Is that for being backward compatible? Don't we have a more robust way to keep the information about original in the future?

Depends what you mean by 'original'. In a recent PR we introduced the field "original_id" to store a reference to the original image when creating a derivative (resize/crop/changing the quality), and this PR changes the crop system to use this instead of the url field on the attachment, which was causing many problems.

If by original you mean the attachment that corresponds to a given url, then no, we currently don't really have a better way, because the DOM is both our presentation layer and data layer.

sdegueldre commented:
@qsm-odoo cleaned things up for a first review, please take a look :)

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_crop_widget.js:
```diff
@@ -1,174 +1,238 @@
-        var src = this.$media.attr('src').split('?')[0];
-        this.aspectRatioList = [
-            [_t("Free"), '0/0', 0],
-            ["16:9", '16/9', 16 / 9],
```
Feels like the logic of the second value is gone in your PR. You are in your web comparing a float parsed from a data attribute with a computed value. I am pretty sure that `parseFloat('1.33333333333333') === (4/3)` is false for any acceptable number of threes to save in the DOM ;)

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_crop_widget.js:
```diff
@@ -1,174 +1,238 @@
+        this.initialSrc = src;
+        // This part of the crop state is not inside cropperData so we need to maintain it manually
+        this.aspectRatio = parseFloat(data.aspectRatio) || 0;
+        this.mimetype = data.mimetype || (_.str.endsWith(src, '.png') ? 'image/png' : 'image/jpeg');
```
Note: you can use `endsWith` now, you are touching the line, you become responsible :D

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_crop_widget.js:
```diff
@@ -1,174 +1,238 @@
-            key = _.str.dasherize(key);
-            self.$media.attr('data-' + key, value);
-            self.$media.data(key, value);
+    _save() {
```
Don't forget to move that in a private section if this becomes private.

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_crop_widget.js:
```diff
@@ -1,174 +1,238 @@
+     *
+     * @private
+     */
+    _discard() {
+        this.media.setAttribute('src', this.initialSrc);
+        return this.destroy();
+    },
```
Feels like this should simply be an override of the destroy method. Maybe the `else` part of your condition in there. If the parent of the widget is destroyed, the widget will be destroyed and the initialSrc should be restored, right?

sdegueldre commented on addons/web_editor/static/src/js/editor/rte.summernote.js:
```diff
@@ -1092,33 +1092,32 @@ var SummernoteManager = Class.extend(mixins.EventDispatcherMixin, ServicesMixin,
 
-            var datas = $croppedImg.attr('src').split(',')[1];
             let attachmentID = cropID;
+            var params = {
```
The first commit is a fix for 12.0 and this line comes from that. see odoo/odoo#49375

sdegueldre commented on addons/web_editor/controllers/main.py:
```diff
@@ -155,17 +155,17 @@ def _update_checklist_recursive (self, li, checked, children=False, ancestors=Fa
         return True
 
     @http.route('/web_editor/attachment/add_data', type='json', auth='user', methods=['POST'], website=True)
-    def add_data(self, name, data, quality=0, width=0, height=0, res_id=False, res_model='ir.ui.view', filters=False, **kwargs):
```
Yes, will do in a separate commit then

sdegueldre commented on addons/web_editor/static/src/js/editor/rte.summernote.js:
```diff
@@ -1092,33 +1092,32 @@ var SummernoteManager = Class.extend(mixins.EventDispatcherMixin, ServicesMixin,
+                datas: $croppedImg.attr('src').split(',')[1],
+                mimetype: $croppedImg.data('crop:mimetype'),
+                // TODO: use original id.
+                url: originalSrc, // To save the original image that was cropped
```
Same thing as above: odoo/odoo#49375

sdegueldre commented on addons/web_editor/controllers/main.py:
```diff
@@ -225,27 +225,34 @@ def remove(self, ids, **kwargs):
+        id_match = re.search('^/web/image/([^/?]+)', src)
+        attachment = []
+        if id_match:
+            number_match = re.match('^(\d+)', id_match.group(1))
```
Apparently, it is possible. Should I do something about it?

sdegueldre commented:
@qsm-odoo Alright, I've rebased this on master after the wrong-original fix was forward ported, and addressed all your previous points, this should be ready for a proper review :)

qsm-odoo commented on addons/web_editor/static/src/js/wysiwyg/widgets/image_crop_widget.js:
```diff
@@ -1,174 +1,237 @@
+            left: 0;
+            right: 0;
+            z-index: 1024;
+        `;
```
All of this should be done another way:
- Warning and checking for the croppability (yes :)) of the image should not be the role of the widget. Instead, you can create an external function (or a static function maybe) whose role is to instantiate the widget *after checking the cropability*. In the case it is not croppable, the function can throw an error with the message you chose. Then the function caller can decide whether or not it wants to catch that particular error type and if it wants to display a warning in that case.
- What is the `$el` of this widget ? Is this this backdrop? In any case, I feel there should be a template definiton for this widget and use of xml and classes instead of javascript and inline style.

sdegueldre replied:
recapping from discord for completeness:

The thing is, checking for croppability requires looking up the image in the database. So should I externalized that as well?

> IMO, the widget should only be responsible for the cropping, it is given a croppable image as input and it outputs a cropped image. Even the saving should probably be externalized.

---

Sorry but I'm not following. Taking a croppable image as input and outputting a cropped image is what the cropper library does. This widget checks for croppability, aligns the cropper with the image, and saves the crop data at the end. Basically you're telling me to move all of what this widget does to the the caller?

qsm-odoo replied:
Ok then the question is more simple: why is it a widget if it is only something which calls a library and makes calls to the server?

sdegueldre replied:
Because it has UI elements that the library doesn't provide (the backdrop and the buttons). I could split this into a Widget for the UI stuff and a bunch of utility functions for loading/storing crop data and attachment data, I'm just not sure why that's desirable.

qsm-odoo replied:
If you use the $el / template properties then it makes sense to use a widget. What made me doubt is that there was none of that when I reviewed or I missed it. Anyways, it seems to be the case now, so let's keep thing that way. It's still not perfect to have a widget/class which instantiates... then only decide to destroy itself if it should not have been instantiated. But let's no waste time on that then.

sdegueldre commented on addons/web_editor/static/src/js/editor/rte.summernote.js:
```diff
@@ -1090,44 +1090,24 @@ var SummernoteManager = Class.extend(mixins.EventDispatcherMixin, ServicesMixin,
+                    mimetype: croppedImg.dataset.mimetype,
+                    original_id: parseInt(croppedImg.dataset.originalId),
+                },
+            });
+            croppedImg.setAttribute('src', croppedAttachmentSrc);
+            weWidgets.ImageCropWidget.prototype.removeOnSaveAttributes.forEach(attr => {
+                delete croppedImg.dataset[attr];
             });
-            $croppedImg.attr('src', '/web/image/' + attachmentID + '?access_token=' + access_token[0]);
```
It's now all done in the controller crop_attachment route on the server, so that I can avoid an RPC round-trip to first create the attachment and then query the db for the image_src.

qsm-odoo replied:
I understood that :wink: What I don't see is the access_token generation server-side?

sdegueldre replied:
For some reason I thought that _compute_image_src did that but it does not. I guess I confused access_token and unique. Added something to handle that in the controller as well.

qsm-odoo commented:
Finally, seems better to retest the whole task with crop + opti together, as well as reviewing them together. So continuing on https://github.com/odoo/odoo/pull/51517
